fail to compile on aarch64 linux

I was experimenting with cross-platform compilation and encountered the following error. It appears the issue might be due to the hardcoded cast of `c_char` to `i8`, where `c_char` can be either `i8` or `u8`, depending on the architecture.(https://doc.rust-lang.org/std/os/raw/type.c_char.html).

```
  Compiling tskit v0.15.0-alpha.2
error[E0308]: mismatched types
    --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/edge_table.rs:67:17
     |
  61 |             Ok(tsk_edge_table_add_row(
     |                ---------------------- arguments to this function are incorrect
...
  67 |                 metadata.as_ptr().cast::<i8>(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /data/target/aarch64-unknown-linux-gnu/release/build/tskit-d34e5d8a61359e9a/out/auto_bindings.rs:5554:12
     |
5554 |     pub fn tsk_edge_table_add_row(
     |            ^^^^^^^^^^^^^^^^^^^^^^
...
5560 |         metadata: *const ::std::os::raw::c_char,
     |         --------

error[E0308]: mismatched types
    --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/individual_table.rs:67:17
     |
  60 |             Ok(tsk_individual_table_add_row(
     |                ---------------------------- arguments to this function are incorrect
...
  67 |                 metadata.as_ptr().cast::<i8>(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /data/target/aarch64-unknown-linux-gnu/release/build/tskit-d34e5d8a61359e9a/out/auto_bindings.rs:5238:12
     |
5238 |     pub fn tsk_individual_table_add_row(
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
5245 |         metadata: *const ::std::os::raw::c_char,
     |         --------

error[E0308]: mismatched types
    --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/migration_table.rs:74:17
     |
  66 |             Ok(tsk_migration_table_add_row(
     |                --------------------------- arguments to this function are incorrect
...
  74 |                 metadata.as_ptr().cast::<i8>(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /data/target/aarch64-unknown-linux-gnu/release/build/tskit-d34e5d8a61359e9a/out/auto_bindings.rs:5707:12
     |
5707 |     pub fn tsk_migration_table_add_row(
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
5715 |         metadata: *const ::std::os::raw::c_char,
     |         --------

error[E0308]: mismatched types
  --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/mutation_table.rs:72:32
   |
72 |                     Some(d) => d.as_ptr() as *const i8,
   |                                ^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`

error[E0308]: mismatched types
    --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/mutation_table.rs:79:17
     |
  65 |             Ok(tsk_mutation_table_add_row(
     |                -------------------------- arguments to this function are incorrect
...
  79 |                 metadata.as_ptr().cast::<i8>(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /data/target/aarch64-unknown-linux-gnu/release/build/tskit-d34e5d8a61359e9a/out/auto_bindings.rs:6016:12
     |
6016 |     pub fn tsk_mutation_table_add_row(
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
...
6024 |         metadata: *const ::std::os::raw::c_char,
     |         --------

error[E0308]: mismatched types
    --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/node_table.rs:83:17
     |
  77 |             tsk_node_table_add_row(
     |             ---------------------- arguments to this function are incorrect
...
  83 |                 metadata.as_ptr().cast::<i8>(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /data/target/aarch64-unknown-linux-gnu/release/build/tskit-d34e5d8a61359e9a/out/auto_bindings.rs:5405:12
     |
5405 |     pub fn tsk_node_table_add_row(
     |            ^^^^^^^^^^^^^^^^^^^^^^
...
5411 |         metadata: *const ::std::os::raw::c_char,
     |         --------

error[E0308]: mismatched types
    --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/population_table.rs:47:17
     |
  45 |             Ok(tsk_population_table_add_row(
     |                ---------------------------- arguments to this function are incorrect
  46 |                 self.as_mut(),
  47 |                 metadata.as_ptr().cast::<i8>(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /data/target/aarch64-unknown-linux-gnu/release/build/tskit-d34e5d8a61359e9a/out/auto_bindings.rs:6180:12
     |
6180 |     pub fn tsk_population_table_add_row(
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
6181 |         self_: *mut tsk_population_table_t,
6182 |         metadata: *const ::std::os::raw::c_char,
     |         --------

error[E0308]: mismatched types
  --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/site_table.rs:60:32
   |
60 |                     Some(d) => d.as_ptr() as *const i8,
   |                                ^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
   |
   = note: expected raw pointer `*const u8`
              found raw pointer `*const i8`

error[E0308]: mismatched types
    --> /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tskit-0.15.0-alpha.2/src/sys/site_table.rs:67:17
     |
  56 |             Ok(tsk_site_table_add_row(
     |                ---------------------- arguments to this function are incorrect
...
  67 |                 metadata.as_ptr().cast::<i8>(),
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /data/target/aarch64-unknown-linux-gnu/release/build/tskit-d34e5d8a61359e9a/out/auto_bindings.rs:5864:12
     |
5864 |     pub fn tsk_site_table_add_row(
     |            ^^^^^^^^^^^^^^^^^^^^^^
...
5869 |         metadata: *const ::std::os::raw::c_char,
     |         --------

For more information about this error, try `rustc --explain E0308`.
warning: tskit@0.15.0-alpha.2: zig: error: version '.2.17' in target triple 'aarch64-unknown-linux-gnu.2.17' is invalid
error: could not compile `tskit` (lib) due to 9 previous errors
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `env -u CARGO AR_aarch64_unknown_linux_gnu="/root/.cache/cargo-zigbuild/0.20.1/ar" CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER="/root/.cache/cargo-zigbuild/0.20.1/zigcc-aarch64-unknown-linux-gnu.2.17-f712.sh" CARGO_ZIGBUILD_RUSTC_VERSION="1.92.0" CC_aarch64_unknown_linux_gnu="/root/.cache/cargo-zigbuild/0.20.1/zigcc-aarch64-unknown-linux-gnu.2.17-f712.sh" CMAKE_TOOLCHAIN_FILE_aarch64_unknown_linux_gnu="/root/.cache/cargo-zigbuild/0.20.1/cmake/aarch64-unknown-linux-gnu-toolchain.cmake" CXX_aarch64_unknown_linux_gnu="/root/.cache/cargo-zigbuild/0.20.1/zigcxx-aarch64-unknown-linux-gnu.2.17-f712.sh" PYO3_BUILD_EXTENSION_MODULE="1" PYO3_CONFIG_FILE="/data/target/maturin/pyo3-config-aarch64-unknown-linux-gnu-3.9.txt" RANLIB_aarch64_unknown_linux_gnu="/root/.cache/cargo-zigbuild/0.20.1/zigranlib.sh" ZIG_COMMAND="/usr/bin/python3 -m ziglang" "cargo" "rustc" "--profile" "release" "--target" "aarch64-unknown-linux-gnu" "--message-format" "json-render-diagnostics" "--manifest-path" "/data/Cargo.toml" "--lib"`

```

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -13,7 +13,7 @@ jobs:
     #  image: ubuntu:20.10
     strategy:
       matrix:
-        os: [ubuntu-24.04, macos-latest]
+        os: [ubuntu-24.04, ubuntu-24.04-arm,  macos-latest]
         rust:
           - stable
     steps:
@@ -32,7 +32,7 @@ jobs:
       - uses: Swatinem/rust-cache@v2.8.2
       - uses: taiki-e/install-action@cargo-hack
       - run: sudo apt-get update -y
-        if: matrix.os == 'ubuntu-24.04'
+        if: (matrix.os == 'ubuntu-24.04') || (matrix.os == 'ubuntu-24.04-arm')
       - name: update toolchain
         run: rustup toolchain install
       - name: cargo check (powerset)

--- a/src/metadata.rs
+++ b/src/metadata.rs
@@ -239,13 +239,14 @@ pub use crate::sys::MetadataError;
 #[cfg(test)]
 mod tests {
     use super::*;
+    use std::ffi::c_char;
 
     #[test]
     fn test_vec8_cast_to_c_string() {
         let v: Vec<u8> = vec![0, 1, b'\0', 2, 3];
         let c = v.as_ptr() as *const libc::c_char;
         for (i, vi) in v.iter().enumerate() {
-            assert_eq!(*vi as i8, unsafe { *c.add(i) });
+            assert_eq!(*vi as c_char, unsafe { *c.add(i) });
         }
 
         let _ = match Some(&v) {

--- a/src/sys/edge_table.rs
+++ b/src/sys/edge_table.rs
@@ -1,3 +1,4 @@
+use std::ffi::c_char;
 use std::ptr::NonNull;
 
 use super::newtypes::EdgeId;
@@ -64,7 +65,7 @@ impl EdgeTable {
                 right,
                 parent,
                 child,
-                metadata.as_ptr().cast::<i8>(),
+                metadata.as_ptr().cast::<c_char>(),
                 metadata.len() as u64,
             ))
         }

--- a/src/sys/individual_table.rs
+++ b/src/sys/individual_table.rs
@@ -1,3 +1,4 @@
+use std::ffi::c_char;
 use std::ptr::NonNull;
 
 use super::bindings::tsk_size_t;
@@ -65,7 +66,7 @@ impl IndividualTable {
                 location.len() as u64,
                 parents.as_ptr(),
                 parents.len() as u64,
-                metadata.as_ptr().cast::<i8>(),
+                metadata.as_ptr().cast::<c_char>(),
                 metadata.len() as u64,
             ))
         }

--- a/src/sys/migration_table.rs
+++ b/src/sys/migration_table.rs
@@ -1,3 +1,4 @@
+use std::ffi::c_char;
 use std::ptr::NonNull;
 
 use super::newtypes::MigrationId;
@@ -71,7 +72,7 @@ impl MigrationTable {
                 source,
                 dest,
                 time,
-                metadata.as_ptr().cast::<i8>(),
+                metadata.as_ptr().cast::<c_char>(),
                 metadata.len() as u64,
             ))
         }

--- a/src/sys/mutation_table.rs
+++ b/src/sys/mutation_table.rs
@@ -1,3 +1,4 @@
+use std::ffi::c_char;
 use std::ptr::NonNull;
 
 use super::newtypes::MutationId;
@@ -70,14 +71,14 @@ impl MutationTable {
                 parent,
                 time,
                 match derived_state {
-                    Some(d) => d.as_ptr() as *const i8,
+                    Some(d) => d.as_ptr() as *const c_char,
                     None => std::ptr::null(),
                 },
                 match derived_state {
                     Some(d) => d.len() as u64,
                     None => 0,
                 },
-                metadata.as_ptr().cast::<i8>(),
+                metadata.as_ptr().cast::<c_char>(),
                 metadata.len() as u64,
             ))
         }

--- a/src/sys/node_table.rs
+++ b/src/sys/node_table.rs
@@ -1,3 +1,4 @@
+use std::ffi::c_char;
 use std::ptr::NonNull;
 
 use super::flags::NodeFlags;
@@ -80,7 +81,7 @@ impl NodeTable {
                 time.into().into(),
                 population.into().into(),
                 individual.into().into(),
-                metadata.as_ptr().cast::<i8>(),
+                metadata.as_ptr().cast::<c_char>(),
                 metadata.len() as u64,
             )
         } {

--- a/src/sys/population_table.rs
+++ b/src/sys/population_table.rs
@@ -1,3 +1,4 @@
+use std::ffi::c_char;
 use std::ptr::NonNull;
 
 use super::bindings::tsk_id_t;
@@ -44,7 +45,7 @@ impl PopulationTable {
         unsafe {
             Ok(tsk_population_table_add_row(
                 self.as_mut(),
-                metadata.as_ptr().cast::<i8>(),
+                metadata.as_ptr().cast::<c_char>(),
                 metadata.len() as u64,
             ))
         }

--- a/src/sys/provenance_table.rs
+++ b/src/sys/provenance_table.rs
@@ -10,6 +10,7 @@ use super::bindings::tsk_provenance_table_t;
 use super::bindings::tsk_size_t;
 use super::tskbox::TskBox;
 use super::TskitError;
+use std::ffi::c_char;
 
 #[derive(Debug)]
 pub struct ProvenanceTable(TskBox<tsk_provenance_table_t>);
@@ -49,9 +50,9 @@ impl ProvenanceTable {
         let rv = unsafe {
             tsk_provenance_table_add_row(
                 self.as_mut(),
-                timestamp.as_ptr() as *mut i8,
+                timestamp.as_ptr() as *const c_char,
                 timestamp.len() as tsk_size_t,
-                record.as_ptr() as *mut i8,
+                record.as_ptr() as *const c_char,
                 record.len() as tsk_size_t,
             )
         };

--- a/src/sys/site_table.rs
+++ b/src/sys/site_table.rs
@@ -1,3 +1,4 @@
+use std::ffi::c_char;
 use std::ptr::NonNull;
 
 use super::bindings::tsk_size_t;
@@ -58,14 +59,14 @@ impl SiteTable {
                 self.as_mut(),
                 position,
                 match ancestral_state {
-                    Some(d) => d.as_ptr() as *const i8,
+                    Some(d) => d.as_ptr() as *const c_char,
                     None => std::ptr::null(),
                 },
                 match ancestral_state {
                     Some(d) => d.len() as u64,
                     None => 0,
                 },
-                metadata.as_ptr().cast::<i8>(),
+                metadata.as_ptr().cast::<c_char>(),
                 metadata.len() as u64,
             ))
         }

--- a/src/trees/treeseq.rs
+++ b/src/trees/treeseq.rs
@@ -18,6 +18,9 @@ use crate::TreeSequenceFlags;
 use crate::TskReturnValue;
 use sys::bindings as ll_bindings;
 
+#[cfg(feature = "provenance")]
+use std::ffi::c_char;
+
 use super::Tree;
 
 /// A tree sequence.
@@ -468,9 +471,9 @@ impl TreeSequence {
         let rv = unsafe {
             ll_bindings::tsk_provenance_table_add_row(
                 &mut (*self.inner.as_ref().tables).provenances,
-                timestamp.as_ptr() as *mut i8,
+                timestamp.as_ptr() as *const c_char,
                 timestamp.len() as ll_bindings::tsk_size_t,
-                record.as_ptr() as *mut i8,
+                record.as_ptr() as *const c_char,
                 record.len() as ll_bindings::tsk_size_t,
             )
         };

